examples/lock-app/cc13x2x7_26x2x7/src/LockManager.cpp

```diff
@@ -0,0 +1,703 @@
+
+    VerifyOrReturnValue(IsValidUserIndex(userIndex), false);
+
+    auto & userInStorage = mLockUsers[userIndex];
+
+    if (userName.size() > DOOR_LOCK_MAX_USER_NAME_SIZE)
+    {
+        ChipLogError(Zcl, "Cannot set user - user name is too long [endpoint=%d,index=%d]", endpointId, userIndex);
+        return false;
+    }
+
+    if (totalCredentials > LockParams.numberOfCredentialsPerUser)
+    {
+        ChipLogError(Zcl, "Cannot set user - total number of credentials is too big [endpoint=%d,index=%d,totalCredentials=%u]",
+                     endpointId, userIndex, totalCredentials);
+        return false;
+    }
+
+    chip::Platform::CopyString(mUserNames[userIndex], userName);
+    userInStorage.userName       = chip::CharSpan(mUserNames[userIndex], userName.size());
+    userInStorage.userUniqueId   = uniqueId;
+    userInStorage.userStatus     = userStatus;
+    userInStorage.userType       = usertype;
+    userInStorage.credentialRule = credentialRule;
+    userInStorage.lastModifiedBy = modifier;
+    userInStorage.createdBy      = creator;
+
+    for (size_t i = 0; i < totalCredentials; ++i)
+    {
+        mCredentials[userIndex][i]                 = credentials[i];
+        mCredentials[userIndex][i].CredentialType  = 1;
+        mCredentials[userIndex][i].CredentialIndex = i + 1;
+    }
+
+    userInStorage.credentials = chip::Span<const DlCredential>(mCredentials[userIndex], totalCredentials);
+
+    // Save user information in NVM flash
+    KeyValueStoreMgr().Put(LockUser, reinterpret_cast<const void *>(&mLockUsers),
+                           sizeof(EmberAfPluginDoorLockUserInfo) * LockParams.numberOfUsers);
+
+    KeyValueStoreMgr().Put(UserCredentials, reinterpret_cast<const void *>(mCredentials),
+                           sizeof(DlCredential) * LockParams.numberOfUsers * LockParams.numberOfCredentialsPerUser);
+
+    KeyValueStoreMgr().Put(LockUserName, reinterpret_cast<const void *>(mUserNames), sizeof(mUserNames));
+
+    ChipLogProgress(Zcl, "Successfully set the user [mEndpointId=%d,index=%d]", endpointId, userIndex);
+
+    return true;
+}
+
+bool LockManager::GetCredential(chip::EndpointId endpointId, uint16_t credentialIndex, DlCredentialType credentialType,
+                                EmberAfPluginDoorLockCredentialInfo & credential)
+{
+
+    if (DlCredentialType::kProgrammingPIN == credentialType)
+    {
+        VerifyOrReturnValue(IsValidCredentialIndex(credentialIndex, credentialType),
+                            false); // programming pin index is only index allowed to contain 0
+    }
+    else
+    {
+        VerifyOrReturnValue(IsValidCredentialIndex(--credentialIndex, credentialType), false); // otherwise, indices are one-indexed
+    }
+
+    ChipLogProgress(Zcl, "Lock App: LockManager::GetCredential [credentialType=%u], credentialIndex=%d",
+                    to_underlying(credentialType), credentialIndex);
+
+    const auto & credentialInStorage = mLockCredentials[credentialIndex];
```

Comment:
This is not spec-compliant.  The index is supposed to be per-type.

Nothing in the test plans tests this, which is why locks that support more than one credential type are provisional, but we should really try to get this right in at least our example apps...